### DISCUSSION TITLE
Connecting ESP32 with AWS IoT Core using MQTT

### DISCUSSION THREAD

**Discussion by nickthiru** [Moddable-OpenSource/moddable#961] [Q&A] [answered]
Hello,

I am trying to connect an ESP32 with AWS IoT Core.

It is my understanding that to connect successfully with AWS IoT Core and, specifically, to the MQTT Broker, I will need to use an MQTT client, a secure communication using TLS 1.2, and provision for (e.g. via AWS IoT Core Console) and use a X.509 root CA certificate, a device certificate, and a private key, all with the necessary AWS IoT policies set. My doubts revolve around where and how to set the two certificates and private key.

Based on the information presented in chapter 3 (Networking) of the "IoT Development for ESP32 and ESP8266 with JavaScript" book, I believe that the root CA certificate needs to be linked by adding it to the resources property in the `manifest.json` string.

With regards to the device certificate, would that be considered a private certificate? If so, do I follow the instructions in the "Private Certificates" section of the same chapter? Meaning, 1) I should add the device certificate to the `resources` property of the `manifest.json` string, 2) create a `Resource` object in the main program, passing the name of the resource file to the Resource constructor, and 3) setting the `secure` property on the connection object that is passed to the MQTT constructor.

Finally, do I add the private key in the same manner as the device certificate (as assumed above)? If so, how would I specify the private key resource in the `secure` property object of the connection object that is passed to the MQTT constructor? In the book's example, only one certificate is attached:
![image](https://user-images.githubusercontent.com/47299208/201257984-49fa821b-1533-4666-84e0-bbc6b479010d.png)

Thank you for you kind assistance.

**Comment by phoddie** [comment, marked as answer]
It is has been a little while since I've brought-up an AWS client. The configuration is complex because you need both the AWS site certificate and the client certificate pieces. Here's an example of making a secure MQTT connection to AWS with a device key & certificate.

```js
new MQTT({
	host,
	id,
	port: 8883,
	Socket: SecureSocket,
	secure: {
		cache: false,
		protocolVersion: 0x303,
		certificate: new Resource("AmazonRootCA1.der"),
		clientKey: new Resource("device.key.der"),
		clientCertificates: [new Resource("device.crt.a.der"), new Resource("device.crt.b.der")],
		verify: true,
		trace: true
	},
});
```

Note: This example shows two client certificates for completeness. However, in most situations you will have one.

Note:  all the certificates and keys need ot be in DER format. If you have them in PEM format, they need to be converted to DER. That can be done with openssl. (The Moddable SDK contains a [function to convert PEM to DER](https://github.com/Moddable-OpenSource/moddable/blob/public/documentation/crypt/crypt.md#class-transform) but an off-line conversion is preferred whenever possible).

**Reply by nickthiru** [reply]
@phoddie Thank you very much for the prompt and helpful response. I will look further into the issue using the details you have provided.

In the AWS IoT developer resources, I came across the following statement: AWS IoT requires devices to send the Server Name Indication (SNI) extension (https://tools.ietf.org/html/rfc3546#section-3.1) when they connect. Does this have any relationship to the implementation in Moddable? For example, is the `id` property in the connection object the SNI that AWS IoT requires?

Thanks again!

**Reply by phoddie** [reply]
The Server Name Indication is sent automatically based on the host. You can verify  that using the simulator and Wireshark.

**Reply by nickthiru** [reply]
Noted. Thanks.

**Reply by nickthiru** [reply]
@phoddie Is there any resource that I might be able to refer to to learn more about the MQTT API and its parameters, like the ones you have showed in your implementation?

**Reply by phoddie** [reply]
The example is mostly the secure [options](https://github.com/Moddable-OpenSource/moddable/blob/public/documentation/network/securesocket.md#constructordictionary) passed through to SecureSocket to configure the TLS connection. Unfortunately, that documentation is incomplete and doesn't cover the client certificate properties.

**Reply by nickthiru** [reply]
Noted. Thanks.

**Reply by phoddie** [reply]
I took a little time today to update the `SecureSocket` documentation:
- New section on [secure MQTT connections](https://github.com/Moddable-OpenSource/moddable/blob/public/documentation/network/securesocket.md#mqtts). That includes the `applicationLayerProtocolNegotiation` property which I seem to remember is sometimes required for AWS IoT.
- Added the `clientCertificates` and `clientKey` properties to the [constructor dictionary table](https://github.com/Moddable-OpenSource/moddable/blob/public/documentation/network/securesocket.md#constructordictionary)
- Added section on [converting PEM to DER](https://github.com/Moddable-OpenSource/moddable/blob/public/documentation/network/securesocket.md#converting-pem-to-der)

Even if you don't need this information now, it is good to finally have it written down for others in the future.

**Reply by nickthiru** [reply]
Awesome! Thank you; greatly appreciated.

I will doing additional research on this very soon, but I just wanted to get your thoughts/approach on the best way to store the certificates and, especially, the private key, in a production environment. A lot of the examples out there, especially in the courses I have been taking, have the certificates and private key in plain text, in the main program. I think I read that the certificates and private key need to be in the same directory as the "manifest.json" file (is this necessary, or can I specify the relative path when calling the Resource constructor?), but should they be stored in some sort of encrypted format? If so, may I know how something like that would work (any resources you can point me to for research would be very welcome :-))?

**Reply by phoddie** [reply]
Good question. There are options depending on the requirements. It is kind of a different topic. Would you mind opening a new discussion for that to help keep thing organized here?

**Reply by nickthiru** [reply]
Good idea. Will do, thank you.
